File: 264_ugly_number_ii/Solution.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [1]
        a = b = c = 0
        for i in range(1, n):
            n2, n3, n5 = dp[a] * 2, dp[b] * 3, dp[c] * 5
            dp.append(min(n2, n3, n5))
            if dp[i] == n2: a += 1
            if dp[i] == n3: b += 1
            if dp[i] == n5: c += 1
        return dp[n - 1]


if __name__ == '__main__':
    print(Solution().nthUglyNumber(10))

File: 264_ugly_number_ii/Solution.go
package main

import "fmt"

func nthUglyNumber(n int) int {
    dp := []int{1}
    a, b, c := 0, 0, 0
    for i := 1; i < n; i++ {
        n2, n3, n5 := dp[a] * 2, dp[b] * 3, dp[c] * 5
        dp = append(dp, min(min(n2, n3), n5))
        if dp[i] == n2 { a++ }
        if dp[i] == n3 { b++ }
        if dp[i] == n5 { c++ }
    }

    return dp[n - 1]
}

func min(a, b int) int {
    if a < b { return a }
    return b
}

func main() {
    fmt.Println(nthUglyNumber(10))
}
